### PR TITLE
Introduce new external endpoints controller

### PR DESCRIPTION
**Pull request by mateiidavid**
For mesh expansion, we need to register an ExternalWorkload's service membership. Service memberships describe which Service objects an ExternalWorkload is part of (i.e. which service can be used to route traffic to an external endpoint).

Service membership will allow the control plane to discover configuration associated with an external endpoint when performing discovery on a service target.

To build these memberships, we introduce a new controller to the destination service, responsible for watching Service and ExternalWorkload objects, and for writing out EndpointSlice objects for each Service that selects one or more external endpoints.

As a first step, we add a new ExternalEndpointManager in the destination service's package that watches services and workloads. In a follow-up change, the ExternalEndpointManager will additionally perform the necessary reconciliation by writing EndpointSlice objects.

Since Linkerd's control plane may run in HA, we also add a lease object that will be used by the manager. When a lease is claimed, a flag is turned on in the manager to let it know it may perform writes.

A more compact list of changes:
* Add a new ExternalEndpointManager file and struct that wires up the necessary mechanisms to watch resources.
  * We also add a new test file; the tests are rudimentary for now.
* Add RBAC rules to the destination service:
  * Allow policy and destination to read ExternalWorkload objects
  * Allow destination to create / update / read Lease objects
* Wire-up mock clients for ExternalWorkload objects in anticipation of more involved tests
---

### Note to reviewers

You can run the controller locally to test the behaviour. You need to make a small change to `controller/destination/main`

* Add `k8s.ExtWorkload` to informers configured (make sure it's the branch that uses endpointslices)
* Create and start the endpoints controller.

```
diff --git a/controller/cmd/destination/main.go b/controller/cmd/destination/main.go
index 845c9e03c..3090e1401 100644
--- a/controller/cmd/destination/main.go
+++ b/controller/cmd/destination/main.go
@@ -9,6 +9,7 @@ import (
        "syscall"

        "github.com/linkerd/linkerd2/controller/api/destination"
+       externalworkload "github.com/linkerd/linkerd2/controller/api/destination/external-workload"
        "github.com/linkerd/linkerd2/controller/api/destination/watcher"
        "github.com/linkerd/linkerd2/controller/k8s"
        "github.com/linkerd/linkerd2/pkg/admin"
@@ -107,7 +108,7 @@ func Main(args []string) {
                        *kubeConfigPath,
                        true,
                        "local",
-                       k8s.Endpoint, k8s.ES, k8s.Pod, k8s.Svc, k8s.SP, k8s.Job, k8s.Srv,
+                       k8s.Endpoint, k8s.ES, k8s.Pod, k8s.Svc, k8s.SP, k8s.Job, k8s.Srv, k8s.ExtWorkload,
                )
        } else {
                k8sAPI, err = k8s.InitializeAPI(
@@ -159,6 +160,9 @@ func Main(args []string) {
        metadataAPI.Sync(nil)
        clusterStore.Sync(nil)

+       ec, _ := externalworkload.NewEndpointsController(k8sAPI, "my-hostname", "default", done)
+       ec.Start()
+
```

You can then run it with `kubectl proxy` (as a bg job) and `go run controller/cmd/main.go destination`. You should see logs from the new `external-endpoints-controller` component.

```
INFO[2024-01-16T14:38:53Z] attempting to acquire leader lease default/linkerd-destination-endpoint-write...
INFO[2024-01-16T14:38:53Z] received default/kubernetes                   component=external-endpoints-controller
INFO[2024-01-16T14:38:53Z] received default/nginx-svc                    component=external-endpoints-controller
```

Sample manifests:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx-svc
spec:
  selector:
    app: nginx
  ports:
    - name: http
      protocol: TCP
      port: 80
---
 apiVersion: workload.linkerd.io/v1alpha1
 kind: ExternalWorkload
 metadata:
  name: test
  labels:
    app: nginx
 spec:
  meshTls:
    serverName: "foo"
    identity: "foo"
  workloadIPs:
  - ip: "192.0.2.0"
  ports:
  - port: 80
```

### CODE REVIEW

File: controller/api/destination/external_endpoint_manager.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+				return
+			}
+		}
+	}()
+}
+
+// enqueueUpdate will enqueue a service key to the updates buffer to be
+// processed by the endpoint manager. When a write to the buffer blocks, the
+// update is dropped.
+func (em *EndpointManager) enqueueUpdate(update string) {
```
Comment (zaharidichev) [inline controller/api/destination/external_endpoint_manager.go:197]: Are we going to introduce some more structured representation to this update or we are going to keep it as a string?

File: controller/api/destination/external_endpoint_manager.go [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+	select {
+	case em.updates <- update:
+		// Update successfully enqueued.
+	default:
+		// Drop update
+		// TODO (matei): add overflow counter
+		em.log.Debugf("External endpoint manager queue is full; dropping update for %s", update)
+		return
```
Comment (zaharidichev) [inline controller/api/destination/external_endpoint_manager.go:205]: You mentioned that you have concerns about simply dropping the update on the floor. Let's think about that (I might be totally wrong). The whole machinery that is present in the destinations service is there because the sink of our updates is a gRPC stream that gets into a state where there is a pending write that blocks the whole pipeline and that **cannot be canceled and will not timeout.**. This last bit is the main reason why the queue there might backup. In my mind, we can avoid getting into a similar state here by: 

- ensuring there is a timeout on the k8s api client that will fail the update operation in case it takes too long
- ensuring the size of the queue is large enough to accommodate a service update (we need to do some back-of-the-envelope math to figure that one out).

In addition to that we need to think whether we can prune these updates. What do we care about in a Service in our case? Do we care if annotations change? Probably not. So thinking through what is a real update vs a resync behaviour.

Does that help?

File: controller/api/destination/external_endpoint_manager.go (hunk 1 above)
Comment (mateiidavid) [reply controller/api/destination/external_endpoint_manager.go]: Probably not. We could pass in a pointer to a service though instead of a string. I think this will become a bit clearer once we move on to reconciliation.

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+	leaseRetryPeriod = 2 * time.Second
+)
+
+// EndpointsController reconciles service memberships for ExternalWorkload resources
+// by writing EndpointSlice objects for Services that select one or more
+// external endpoints.
+type EndpointsController struct {
+	k8sAPI   *k8s.API
+	log      *logging.Entry
+	updates  chan string
```
Comment (zaharidichev) [inline controller/api/destination/external-workload/endpoints_controller.go:44]: I will spent a bit more time thinking about the best way to do that but it would be very useful to have: 

- metrics on current queue size
- metrics on avg time spent in the queue
- overflow counter

Comment (mateiidavid) [reply controller/api/destination/external-workload/endpoints_controller.go]: Added some metrics for the queue. I re-shuffled bits of code around to capture some of the data (e.g. time spent in the queue). Let me know if this looks okay to you.

We're basically computing:
1. Histograms for queue latency (time spent in queue, the buckets were a bit arbitrary would love some feedback there)
2. Number of total elements enqueued
3. Queue length
4. Number of elements dropped from the queue (I thought this might make more sense than an overflow. For the translator we have an overflow counter per translator, once it overflows it kills the stream, we continue to work here so it felt better to capture how many updates we've shed).

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+			if old.ResourceVersion == updated.ResourceVersion {
+				ec.log.Tracef("skipping ExternalWorkload resync update event")
+				return
+			}
+
+			// Compute whether any labels have changed in between updates. If
+			// they have, this might affect service membership so we need to
+			// process the update.
+			labelsChanged := false
+			if len(old.Labels) != len(updated.Labels) {
```
Comment (adleong) [inline controller/api/destination/external-workload/endpoints_controller.go:214]: TIOLI style: I like putting functionality like this into a helper function (e.g. `func labelsChanged(old, updated Labels) bool`) so that I can take advantage of early return instead of using a boolean var.

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+				}
+			}
+
+			// If nothing has changed, don't bother with the update
+			if !readinessChanged && !specChanged && !labelsChanged {
+				ec.log.Debugf("skipping ExternalWorkload update; nothing has changed between old rv %s and new rv %s", old.ResourceVersion, updated.ResourceVersion)
+				return
+			}
+
+			services, err := ec.getSvcMembership(updated)
```
Comment (adleong) [inline controller/api/destination/external-workload/endpoints_controller.go:261]: Do we also have to recompute membership in services that old is part of?  In case this workload is no longer part of a service?

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+			}
+
+			// Ignore resync updates. If nothing has changed in the object, then
+			// the update processing is redudant.
+			if old.ResourceVersion == updated.ResourceVersion {
+				ec.log.Tracef("skipping ExternalWorkload resync update event")
+				return
+			}
+
+			// Compute whether any labels have changed in between updates. If
```
Comment (adleong) [inline controller/api/destination/external-workload/endpoints_controller.go:210]: There's a lot of logic here to try to determine if the change to the ExternalWorkload is a material difference (and therefore if endpoints should be recomputed for any services).  It would simplify things if we enqueued the service updates for any change here.  Do you think this would cause problematic load? It would only trigger updates for services which select the old or new ExternalWorkload.

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+			if !ok {
+				ec.log.Errorf("error processing ExternalWorkload event: expected *v1alpha1.ExternalWorkload, got %#v", oldObj)
+			}
+
+			updated, ok := newObj.(*ewv1alpha1.ExternalWorkload)
+			if !ok {
+				ec.log.Errorf("error processing ExternalWorkload event: expected *v1alpha1.ExternalWorkload, got %#v", newObj)
+			}
+
+			// Ignore resync updates. If nothing has changed in the object, then
```
Comment (adleong) [inline controller/api/destination/external-workload/endpoints_controller.go:203]: Are you sure we want to ignore resyncs?  What if we miss an update, how will we recover?  E.g. suppose writing the EndpointSlice for a service fails and therefore an update to an ExternalWorkload is not reflected in the EndpointSlice.  How and when does the state get corrected?

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+	// Start a background task to process updates. When a shutdown signal is
+	// received over the manager's stop channel, it is propagated to the elector
+	// through the context object to ensure the elector task does not leak.
+	go func() {
+		for {
+			select {
+			case update := <-ec.updates.queue:
+				elapsed := time.Since(update.enqueTime).Seconds()
+				ec.updates.queueLatency.Observe(elapsed)
+				ec.updates.queueLength.Dec()
```
Comment (adleong) [inline controller/api/destination/external-workload/endpoints_controller.go:397]: Can we register a gauge function that returns `len(ec.updates.queue)` instead of managing the gauge value manually?

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 9]
```diff
@@ -0,0 +1,513 @@
...
+	// TODO (matei): reconciliation logic
+	// TODO (matei): track how long an item takes to be processed
+	ec.log.Infof("Received %s", update)
+}
+
+// getSvcMembership accepts a pointer to an external workload resource and
+// returns a set of service keys (<namespace>/<name>). The set includes all
+// services local to the workload's namespace that match the workload.
+func (ec *EndpointsController) getSvcMembership(workload *ewv1alpha1.ExternalWorkload) (map[string]struct{}, error) {
+	set := make(map[string]struct{})
```
Comment (adleong) [inline controller/api/destination/external-workload/endpoints_controller.go:441]: I love sets but in this case where we won't potentially have duplicate keys and where we don't have to check membership, a slice seems more appropriate.

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 10]
```diff
@@ -0,0 +1,513 @@
...
+
+// EndpointsController reconciles service memberships for ExternalWorkload resources
+// by writing EndpointSlice objects for Services that select one or more
+// external endpoints.
+type EndpointsController struct {
+	k8sAPI   *k8s.API
+	log      *logging.Entry
+	updates  updateQueue
+	stop     chan struct{}
+	isLeader bool
```
Comment (adleong) [inline controller/api/destination/external-workload/endpoints_controller.go:91]: making this an atomic bool is probably more straightforward and less error prone than protecting it with a mutex.

File: controller/api/destination/external-workload/endpoints_controller.go (hunk 8 above)
Comment (zaharidichev) [reply controller/api/destination/external-workload/endpoints_controller.go:397]: I think that is right. Each time you touch the queue, just measure the len

File: controller/api/destination/external-workload/endpoints_controller.go (hunk 7 above)
Comment (zaharidichev) [reply controller/api/destination/external-workload/endpoints_controller.go:203]: @adleong I think this is tricky for a couple of reasons. My initial recommendation when I chatted with @mateiidavid was to prune the updates (determine whether it is a material update as you said earlier). I now see this is all a bit more nuanced. The problem with resyncs is the following in my eyes. 

I am assuming we do not miss events coming from the watcher while the controller is running. What might happen is to error when we actually do the write to the k8s API and miss the update this way. A resync can fix that, but the problem is that on the resyncs you get an Update that has two identical old + new. So if you have failed to do an update where an ExternalWorkload transitions from selecting over a service to one that does not select this service anymore, your next resync will give you an Update that has old and new that do not select over the Service. 

In order to remediate the situation you would have to then go over the endpoints of the slices for all services to check whether there are any that should not be there (e.g. refering to this external workload when it should not).
Correct me if I am wrong but this seems rather excessive. This means that if you have 10 ExternalWorkloads in a namespace and 20 services on each resync you would need to do 200 of these full reconciliations. Does that seem correct? 

So two options in my mind: 

- process resyncs and do this full diffing on the resyncs
- have a routine when the controller starts, that fetches all ExternalWorklaods and services and reconciles the state  (to avoid the situation where you have missed updates while the controller was down). So a full state sync. After that if you do not want to deal with all resyncs but are worried about failing writes, couple the processing logic with a retriable queue with a delay (similar to the multicluster controller). This way you can requeue an update if it fails. 

Happy to hear more opinions on the matter.

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 11]
```diff
@@ -0,0 +1,513 @@
...
+	// TODO (matei): track how long an item takes to be processed
+	ec.log.Infof("Received %s", update)
+}
+
+// getSvcMembership accepts a pointer to an external workload resource and
+// returns a set of service keys (<namespace>/<name>). The set includes all
+// services local to the workload's namespace that match the workload.
+func (ec *EndpointsController) getSvcMembership(workload *ewv1alpha1.ExternalWorkload) (map[string]struct{}, error) {
+	set := make(map[string]struct{})
+	services, err := ec.k8sAPI.Svc().Lister().Services(workload.Namespace).List(labels.Everything())
```
[the remark this answers is not in the document]
Comment (mateiidavid) [reply controller/api/destination/external-workload/endpoints_controller.go:442]: As in pass the labels from the workload? That might not fully work since services and workloads can have a different set of labels. For example:

```
apiVersion: v1
kind: Service
metadata:
  name: my-svc
spec:
  selector:
    app=my-app
---
apiVersion: workload.linkerd.io/v1alpha1
kind: ExternalWorkload
metadata:
  name: my-workload
  labels:
    app=my-app
```

This would be a valid example. Since we end up checking the service's _selector_ we can't restrict our list to only get services that match the workload's labels.

File: controller/api/destination/external-workload/endpoints_controller.go (hunk 10 above)
Comment (mateiidavid) [reply controller/api/destination/external-workload/endpoints_controller.go:91]: great suggestion to simplify this 👍🏻

File: controller/api/destination/external-workload/endpoints_controller.go (hunk 5 above)
Comment (mateiidavid) [reply controller/api/destination/external-workload/endpoints_controller.go]: Yup! Labels changing in between updates is an indication that the membership has to be recomputed for the old service too.

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 12]
```diff
@@ -0,0 +1,531 @@
...
+	}
+
+	return false
+}
+
+// servicesToUpdate accepts pointers to two ExternalWorkload resources used to
+// determine the state of an update. Based on the state of the update and the
+// references to the workload resources, it will collect a set of Services that
+// need to be updated.
+func (ec *EndpointsController) servicesToUpdate(old, updated *ewv1alpha1.ExternalWorkload) []string {
```
Comment (adleong) [inline controller/api/destination/external-workload/endpoints_controller.go:427]: I'm having  a really hard time reasoning through this method and figuring out if it's correct.  lots of different cases and hard to keep track of them all.  can it written in an easier to understand way?

if labels have changed, then membership has changed and we need to recompute services for both the old and new workload
otherwise if only workload has changed then membership has not changed and services for old and new workload should be the same

so doesn't this all amount to the same thing? as long as we're dedupping, shouldn't we always recompute services for both old and new?

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 13]
```diff
@@ -0,0 +1,531 @@
...
+		// Taken from official k8s code, this checks whether a given object has
+		// a deleted state before returning a `namespace/name` key. This is
+		// important since we do not want to consider a service that has been
+		// deleted and is waiting for cache eviction
+		key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(svc)
+		if err != nil {
+			return []string{}, err
+		}
+
+		// Check if service selects our ee.
```
Comment (adleong) [inline controller/api/destination/external-workload/endpoints_controller.go:524]: typo?

File: controller/api/destination/external-workload/endpoints_controller.go (hunk 12 above)
Comment (zaharidichev) [reply controller/api/destination/external-workload/endpoints_controller.go:427]: @adleong @mateiidavid  I am having a bit of a hard time following myself. This is why I tried to rewrite it in a different way and added some comments. Let me all know whether this is correct and whether it looks simpler. Also. please make sure to write tests of it that exercise all the edge cases. 

```golang
func toSet(s []string) map[string]struct{} {
	set := map[string]struct{}{}
	for _, k := range s {
		set[k] = struct{}{}
	}
	return set
}

// servicesToUpdate accepts pointers to two ExternalWorkload resources used to
// determine the state of an update. Based on the state of the update and the
// references to the workload resources, it will collect a set of Services that
// need to be updated.
func (ec *EndpointsController) servicesToUpdate(old, updated *ewv1alpha1.ExternalWorkload) []string {
	labelsChanged := labelsChanged(old, updated)
	workloadChanged := workloadChanged(old, updated)
	if !labelsChanged && !workloadChanged {
		ec.log.Debugf("skipping ExternalWorkload update; nothing has changed between old rv %s and new rv %s", old.ResourceVersion, updated.ResourceVersion)
		return nil
	}

	newSelection, err := ec.getSvcMembership(updated)
	if err != nil {
		ec.log.Errorf("failed to get service membership for workload %s/%s: %v", updated.Namespace, updated.Name, err)
		return nil
	}

	oldSelection, err := ec.getSvcMembership(updated)
	if err != nil {
		ec.log.Errorf("failed to get service membership for workload %s/%s: %v", updated.Namespace, updated.Name, err)
		return nil
	}

	result := map[string]struct{}{}
	if labelsChanged && workloadChanged {
		// we need to consider the union of the selection
		// as both selection and workload content
		// has changed
		for _, s := range append(newSelection, oldSelection...) {
			result[s] = struct{}{}
		}
	} else if workloadChanged {
		// workload changed, but selection stays the same
		// in this case newSelection and oldSelection should
		// be the same, so we can safely reconsider all services
		// from newSelection
		return newSelection
	} else {
		// only selection changed, so we need to consider the diff
		// of the new and old selection as the rest of the services
		// should stay the same as content of workload has not changed
		newSelectionSet := toSet(newSelection)
		oldSelectionSet := toSet(oldSelection)

		// consider all services that are in the
		// old selection, but are not in the new
		// one (i.e. removed ones)
		for _, oldSvc := range oldSelection {
			if _, ok := newSelectionSet[oldSvc]; !ok {
				result[oldSvc] = struct{}{}
			}
		}

		// consider all services that are in the
		// new selection, but are not in the old
		// one (i.e. added ones)
		for _, newSvc := range newSelection {
			if _, ok := oldSelectionSet[newSvc]; !ok {
				result[newSvc] = struct{}{}
			}
		}
	}

	var resultSlice []string
	for svc, _ := range result {
		resultSlice = append(resultSlice, svc)
	}
	return resultSlice
}
```

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 14]
```diff
@@ -0,0 +1,531 @@
...
+	}
+
+	if !labelsChanged {
+		return updatedSvc
+	}
+
+	oldSvc, err := ec.getSvcMembership(old)
+	if err != nil {
+		ec.log.Errorf("failed to get service membership for workload %s/%s: %v", old.Namespace, old.Name, err)
+		return nil
```
[the remark this answers is not in the document]
Comment (zaharidichev) [reply controller/api/destination/external-workload/endpoints_controller.go:448]: I agree with that

**Review by zaharidichev — COMMENTED** [review]
Good job. We are getting there. I left some feedback that can hopefully make things easier to read. This diffing logic is always quite error-prone. This is why I think you need to rely on more isolated tests like the one I provided in the comments. This should make it easier to think through and verify stuff.

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 15]
```diff
@@ -0,0 +1,547 @@
...
+
+	_, err = k8sAPI.ExtWorkload().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			ew, ok := obj.(*ewv1alpha1.ExternalWorkload)
+			if !ok {
+				ec.log.Errorf("couldn't get ExternalWorkload from object %#v", obj)
+				return
+			}
+
+			if len(ew.Spec.WorkloadIPs) == 0 {
```
Comment (zaharidichev) [inline controller/api/destination/external-workload/endpoints_controller.go:86]: Hm.. now that I think about it. Here is something you might encounter. We are not relying on resyncs, right. So we rely on the initial Adds once the controller fires up in order to get up to speed. What if we have the following scenario.

1. We have a workload with an IP and a service that selects it.
2. All is good and smooth
3. The controller crashes
4. Someone changes the workload to have zero IPs
5. The controller fires up again
6. We gen an add for this workload and ignore it because it has zero IPs

Can we end up in a situation where we have stale endpoints then? Or will these be resolved because we will get ADD callbacks on the Service informer as well?

**Review by zaharidichev — COMMENTED** [review]
Looks good, one last question to think through.

File: controller/api/destination/external-workload/endpoints_controller.go (hunk 15 above)
Comment (mateiidavid) [reply controller/api/destination/external-workload/endpoints_controller.go:86]: That should be resolved by looking at the service event, yes.

1. Controller crashes.
2. Someone makes changes to the workload to have zero IPs
3. Controller fires up again, does a full re-sync
4. We reconcile the service, get all of its _valid_ workloads
5. Detect a diff between current workload IPs and what's in the endpointslice (will contain IP from changed workload)
6. Remove stale address.

**Review by zaharidichev — APPROVED** [review]
LGTM. Great job.

**Comment by adleong** [comment]
It may be worth noting in a comment that this is structurally based on https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go and perhaps even noting cases where they significantly differ (if any).

File: controller/api/destination/external-workload/endpoints_controller.go [hunk 16]
```diff
@@ -50,6 +50,19 @@ type EndpointsController struct {
 	sync.RWMutex
 }
 
+// The EndpointsController code has been structured (and modified) based on the
+// core EndpointSlice controller. Copyright 2014 The Kubernetes Authors
+// https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go
```
Comment (zaharidichev) [inline controller/api/destination/external-workload/endpoints_controller.go:55]: better point to a specific commit

**Review by zaharidichev — APPROVED** [review]
